SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTest.cpp

```diff
@@ -0,0 +1,387 @@
+				{
+					if (VerifyTestActor(LevelActor, 0, 0, 0, 0))
+					{
+						FinishStep(); // Clients don't have authority over non-replicated Level Actors.
+					}
+				}
+			}, 5.0f);
+	}
+
+	// Replicated Dynamic Actor Spawned On Same Server. Server 1 should have Authority.
+	{
+		AddStep(TEXT("Replicated Dynamic Actor Spawned On Same Server - Spawn"), FWorkerDefinition::Server(1), nullptr,
+			[this]()
+			{
+				ASpatialAuthorityTestReplicatedActor* Actor = GetWorld()->SpawnActor<ASpatialAuthorityTestReplicatedActor>(
+					Server1Position, FRotator::ZeroRotator);
+				CrossServerSetDynamicReplicatedActor(Actor);
+				FinishStep();
+			});
+
+		AddStep(TEXT("Replicated Dynamic Actor Spawned On Same Server - Verify Server 1 Has Authority"), FWorkerDefinition::AllWorkers, nullptr, nullptr,
+			[this](float DeltaTime)
+			{
+				Timer -= DeltaTime;
+				if (Timer <= 0)
+				{
+					const FWorkerDefinition& LocalWorkerDefinition = GetLocalFlowController()->WorkerDefinition;
+					if( LocalWorkerDefinition.Type == ESpatialFunctionalTestWorkerType::Server && LocalWorkerDefinition.Id == 1)
+					{
+						if (VerifyTestActor(DynamicReplicatedActor, 1, 1, 1, 0))
+						{
+							FinishStep();
+						}
+					}
+					else
+					{
+						if (VerifyTestActor(DynamicReplicatedActor, 0, 0, 0, 0))
+						{
+							FinishStep();
+						}
+					}
+				}
+			}, 5.0f);
+
+		AddStep(TEXT("Replicated Dynamic Actor Spawned On Same Server - Destroy"), FWorkerDefinition::Server(1), nullptr,
+			[this]()
+			{
+				DynamicReplicatedActor->Destroy();
+				CrossServerSetDynamicReplicatedActor(nullptr);
+				FinishStep();
+			});
+	}
+
+	// Replicated Dynamic Actor Spawned On Different Server. Server 1 should have Authority on BeginPlay, Server 2 on Tick
+	{
+		AddStep(TEXT("Replicated Dynamic Actor Spawned On Different Server - Spawn"), FWorkerDefinition::Server(1), nullptr,
+				[this]() {
+					ASpatialAuthorityTestReplicatedActor* Actor = GetWorld()->SpawnActor<ASpatialAuthorityTestReplicatedActor>(
+						Server2Position, FRotator::ZeroRotator);
+					CrossServerSetDynamicReplicatedActor(Actor);
+					FinishStep();
+				});
+
+		AddStep(
+			TEXT("Replicated Dynamic Actor Spawned On Different Server - Verify Server 1 Has Authority on BeginPlay and Server 2 on Tick"), FWorkerDefinition::AllWorkers, nullptr,
+			nullptr,
+			[this](float DeltaTime) {
+				Timer -= DeltaTime;
+				if(Timer <= 0)
+				{
+					const FWorkerDefinition& LocalWorkerDefinition = GetLocalFlowController()->WorkerDefinition;
+					if(LocalWorkerDefinition.Type == ESpatialFunctionalTestWorkerType::Server)
+					{
+						// Allow it to continue working in Native / Single worker setups.
+						if (GetNumberOfServerWorkers() > 1)
+						{
+							if (LocalWorkerDefinition.Id == 1)
+							{
+								// Note: Depending on timing / data migration, Tick may or may not have ran
+								// in Server 1, so we need to check both situations.
+								if (VerifyTestActor(DynamicReplicatedActor, 1, 1, 1, 1)
```

Comment:
What's the tick delta for this actor? If it's not 60fps we should increase it and try to remove this edge case.

Reply:
I think this could actually be pointing to a bug - isn't the buffer period supposed to be 1s? I think we should always tick at least once on the original server before migrating - if it does not happen, I suspect that something else may be going wrong?

Reply:
I asked about this in one of the chats and was told that it may or may not happen. Currently in the test I was always seeing at least 1 tick (can't confirm if more), the tick rate is set to 0.0f to ensure ticks every frame (mostly put it like this because like in the case of NWX they override it for actors by default only tick every 1s). The thing that I don't know is how this behaves in multiple situations, like if the machine is running tests without proper resources, low fps, etc.

If it's not meant to happen, then I can change, currently this is the behaviour I'm seeing.

Reply:
Could you link the slack thread?

Reply:
I'll try to find it, was already a week or 2 ago.

Reply:
https://improbable.slack.com/archives/GHKNDKH3K/p1597319216329600

Reply:
Discussed offline, this shouldn't be a race with the current setup, and will be updated accordingly.

Reply:
Changed it to verify that it will always tick before migrating